lib/split/trial.rb

```diff
@@ -3,15 +3,30 @@ module Split
   class Trial
     attr_accessor :experiment
     attr_accessor :metadata
+    attr_accessor :version
 
     def initialize(attrs = {})
       self.experiment   = attrs.delete(:experiment)
       self.alternative  = attrs.delete(:alternative)
       self.metadata  = attrs.delete(:metadata)
+      self.version = attrs.delete(:version)
 
       @user             = attrs.delete(:user)
       @options          = attrs
 
+      if self.version.nil?
+        if @user
+          key_for_experiment = @user.keys.find { |k| k.match(Regexp.new("^#{experiment.name}"))}
+          if key_for_experiment
+            self.version = key_for_experiment.split(":").last
+          else
+            self.version = 0
+          end
+        else
+          self.version = 0
+        end
+      end
```

Comment:
Where are `Split::Trial` objects created? It feels weird that one could be created without know the current version of the experiment it refers to. I'm also wonder if that place is maybe also better place to do the Redis lookup/regex that's happening above. There's an OOP principle that's basically "decouple the logic for creating an object from the object itself". There's a few reason to prefer that decoupling, but concretely if there's another spot that needs to map an experiment name to the "current version" this code can't be used.

Reply:
The concept I was looking for was the Dependency Inversion Principle.

Reply:
I have a difficult time keeping straight the places where a lot of instantiation occurs. Will need to double down on this.

Reply:
I agree that the logic `to fallback to version = 0 in the case that we don't supply a version` should be specified in the calling location where we expect that behavior, instead of here.

Reply:
only trouble is that I have trouble finding all locations, I'm trying to find an exhaustive list.

Reply:
One thing you might try is changing the argument signature of the constructor to something mildly non-sensical.  Say changing:

```ruby
def initialize(attrs = {})

# To

def initialize(foo, bar, attrs = {})
```

Then update the known one or two instantiation points to pass in nil for the first two arguments. If you run the tests + poke on the app I think you'll find them pretty quickly.